src/main/java/learn/petgallery/data/AuditingPetRepository.java

```diff
@@ -0,0 +1,46 @@
+package learn.petgallery.data;
+
+import learn.petgallery.models.EntityEvent;
+import learn.petgallery.models.EntityEventType;
+import learn.petgallery.models.Pet;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class AuditingPetRepository implements PetRepository{
+
+    private final JpaPetRepository jpaPetRepository;
+    private final EntityEventRepository entityEventRepository;
+
+    public AuditingPetRepository(JpaPetRepository jpaPetRepository,
+                                 EntityEventRepository entityEventRepository) {
+        this.jpaPetRepository = jpaPetRepository;
+        this.entityEventRepository = entityEventRepository;
+    }
+
+    @Override
+    public List<Pet> findAll() {
+        return jpaPetRepository.findAll();
+    }
+
+    @Override
+    public Pet findById(int petId) {
+        return jpaPetRepository.findById(petId).orElse(null);
+    }
+
+    @Override
+    public Pet save(Pet pet, int appUserId) {
+        Pet savedPet = jpaPetRepository.save(pet);
+        EntityEvent entityEvent = EntityEvent.createSaveEvent(savedPet, EntityEventType.SAVE, appUserId);
+        entityEventRepository.save(entityEvent);
+        return savedPet;
+    }
+
+    @Override
+    public void deleteById(int petId, int appUserId) {
+        EntityEvent entityEvent = EntityEvent.createDeleteEvent(petId, Pet.class.getName(), appUserId);
+        jpaPetRepository.deleteById(petId);
+        entityEventRepository.save(entityEvent);
+    }
```

Comment:
Add the Transactional annotation to these methods so the event logging and save/delete operations are atomic?

Reply:
Also, would it be possible to create a `JpaAuditingRepository<>` subclass that extends `JpaRepository<>`? The `JpaAuditingRepository<>` subclass could override the `save()` and `deleteById()` methods, adding in the EntityEvent code.

Reply:
I'll explore that idea!

Reply:
I remember, I did explore something along these lines. The difficulty arises with needing `appUserId`, if that is part of the signature then we can't override those methods. We _could_ consider injecting a utility that can pluck the `Principal` out of the security context in the data layer, thereby avoiding changing the signature, but that makes a lot of assumptions about the consuming code.

Reply:
How about adding `save()` and `deleteById()` methods that accept the generic entity type along with the appUserId, that call the base class' `save()` and `deleteById()` methods along with EntityEvent code?